### PR TITLE
Feat/discovery services sync

### PR DESCRIPTION
**Pull request by asulis**
### Description:

This PR implements an automatic or on request sync task fwith Directory. All biobanks and collections are read from directory and compared with organizations and resources in the Negotiator. All the missing are added and all those already present are updated, if necessary.
Work still in progress

### Checklist:

_Make sure you tick all the boxes bellow if they are true or do not apply:_

- [ ] I have performed a self review of my code
- [ ] My code follows [Google Java Code style](https://google.github.io/styleguide/javaguide.html)
- [ ] I have made my code as simple as possible
- [ ] I have added unit tests and the code coverage has not decreased
- [ ] I have updated the documentation in all relevant places

### CODE REVIEW

**Review by RadovanTomik — CHANGES_REQUESTED** [review]
the class and function names are getting ridiculously long. Please put these classes into new packages and rename these classes. The package(s) names should provide context to the new shorter class names.
Remove the new controller and add these methods for fetching and creating sync jobs to the Discovery service controller. E.g. /discovery-services/1/sync-job/... please do not use names to identify objects when we have Ids

File: src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  @Autowired
+  DiscoveryServiceSynchronizationJobRepository discoveryServiceSynchronizationJobRepository;
+
+  @Override
+  public DiscoverySyncJobServiceDTO createSyncJob(
+      DiscoverySyncJobServiceCreateDTO discoverySyncJobServiceCreateDTO) {
+    String discoveryServiceName = discoverySyncJobServiceCreateDTO.getDiscoveryServiceName();
+    Optional<DiscoveryService> discoveryService =
+        discoveryServiceRepository.findByName(discoveryServiceName);
+    if (!discoveryService.isEmpty()) {
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java:34]: just mandate the use of all args constructor

File: src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      LocalDateTime creationDate = LocalDateTime.now();
+      LocalDateTime modifyDate = LocalDateTime.now();
+      DiscoveryServiceSynchronizationJob job =
+          DiscoveryServiceSynchronizationJob.builder()
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java:38]: use mapper

File: src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+              .service(discoveryService.get())
+              .creationDate(creationDate)
+              .modifiedDate(modifyDate)
+              .status(DiscoveryServiceSyncronizationJobStatus.SUBMITTED)
+              .build();
+
+      log.info("Saving new job entity....");
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java:45]: use debug level

File: src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+  @Autowired DiscoveryServiceRepository discoveryServiceRepository;
+
+  @Autowired DiscoveryServiceSynchronizationEventPublisher publisher;
+
+  @Autowired
+  DiscoveryServiceSynchronizationJobRepository discoveryServiceSynchronizationJobRepository;
+
+  @Override
+  public DiscoverySyncJobServiceDTO createSyncJob(
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java:29]: the whole method is really complicated for a simple create

File: src/main/java/eu/bbmri_eric/negotiator/service/MolgenisService.java [hunk 5]
```diff
@@ -27,4 +28,19 @@ public interface MolgenisService {
    * @return Optional<MolgenisBiobank>
    */
   Optional<MolgenisBiobank> findBiobankById(String id);
+
+  /**
+   * Fetch all BBMRI-ERIC, MIABIS defined Biobanks.
+   *
+   * @return Optional List<MolgenisBiobank>
+   */
+  List<MolgenisBiobank> findAllBiobanks();
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/MolgenisService.java:37]: no need to use the MolgenisService, tbh I would delete it. Just use the discovery service client

**Review by RadovanTomik — CHANGES_REQUESTED** [review]
please see comments

File: src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/DiscoverySyncronizationJobController.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/v3")
+@CommonsLog
+@SecurityRequirement(name = "security_auth")
```
[the remark this answers is not in the document]
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/DiscoverySyncronizationJobController.java]: Fixed in 2fe322f311ff3eb6cc2c77cbea7049ff3b94bef2

File: src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/DiscoverySyncronizationJobController.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+@Tag(
+    name = "Discovery service synchronization",
+    description = "Submit and manage discovery service synchronization jobs")
+public class DiscoverySyncronizationJobController {
+
+  @Autowired private DiscoverySynchronizationJobService discoverySynchronizationJobService;
+
+  @PostMapping(
+      value = "/discoveryServiceSyncJob",
```
[the remark this answers is not in the document]
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/api/controller/v3/DiscoverySyncronizationJobController.java]: Fixed in e8071a837f32502b39356d9e2e64164c1c8d3cf8

File: src/main/java/eu/bbmri_eric/negotiator/events/DiscoveryServiceSynchronizationEvent.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import lombok.Setter;
+import org.springframework.context.ApplicationEvent;
+
+/** Event for Discovery Service synchronization * */
+@Getter
+@Setter
+public class DiscoveryServiceSynchronizationEvent extends ApplicationEvent {
+
+  private String jobId;
+  private String discoveryServiceName;
```
[the remark this answers is not in the document]
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/events/DiscoveryServiceSynchronizationEvent.java]: Fixed in e8071a837f32502b39356d9e2e64164c1c8d3cf8

File: src/main/java/eu/bbmri_eric/negotiator/publishers/DiscoveryServiceSynchronizationEventPublisher.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+@CommonsLog
+public class DiscoveryServiceSynchronizationEventPublisher {
+  @Autowired private ApplicationEventPublisher applicationEventPublisher;
+
+  public void publishDiscoveryServiceSynchronizationEvent(
+      final String jobId, final String serviceName) {
+    System.out.println("Publishing custom event. ");
```
[the remark this answers is not in the document]
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/publishers/DiscoveryServiceSynchronizationEventPublisher.java]: Fixed in e8071a837f32502b39356d9e2e64164c1c8d3cf8

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDirectoryServiceSynchClient.java [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package eu.bbmri_eric.negotiator.service;
+
+public interface BBMRIDirectoryServiceSynchClient {
```
[the remark this answers is not in the document]
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDirectoryServiceSynchClient.java]: Fixed in 9be653400f948d151453fe30cf49f1309e4b3ea6

File: src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java (hunk 1 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java]: Fixed in 3a6688e83ae3a2214da370432b86f40523292504

File: src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java (hunk 2 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java]: Fixed in 3a6688e83ae3a2214da370432b86f40523292504

File: src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java (hunk 3 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/DiscoverySynchronizationJobServiceImpl.java]: Fixed in 3a6688e83ae3a2214da370432b86f40523292504

File: src/main/java/eu/bbmri_eric/negotiator/database/model/DiscoveryServiceSynchronizationJob.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+import lombok.ToString;
+import org.hibernate.annotations.UuidGenerator;
+
+@ToString
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter
+@Setter
+@Builder
+@Entity(name = "DiscoveryServiceSynchronizationJob")
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/database/model/DiscoveryServiceSynchronizationJob.java:26]: is this necessary?

File: src/main/java/eu/bbmri_eric/negotiator/database/model/DiscoveryServiceSynchronizationJob.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  @Column(name = "id")
+  private String id;
+
+  @ManyToOne
+  @JoinColumn(name = "discovery_service_id", referencedColumnName = "id")
+  private DiscoveryService service;
+
+  private LocalDateTime creationDate;
+
+  private LocalDateTime modifiedDate;
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/database/model/DiscoveryServiceSynchronizationJob.java:41]: should be finished date imho

File: src/main/java/eu/bbmri_eric/negotiator/database/repository/DiscoveryServiceRepository.java [hunk 13]
```diff
@@ -9,4 +9,6 @@
 public interface DiscoveryServiceRepository extends JpaRepository<DiscoveryService, Long> {
 
   Optional<DiscoveryService> findByUrl(String url);
+
+  Optional<DiscoveryService> findByName(String name);
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/database/repository/DiscoveryServiceRepository.java:13]: is this method necessary?

File: src/main/java/eu/bbmri_eric/negotiator/dto/syncjobservice/DiscoverySyncJobServiceCreateDTO.java [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+import lombok.Setter;
+
+@Getter
+@Setter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class DiscoverySyncJobServiceCreateDTO {
+
+  @NotNull private String discoveryServiceName;
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/dto/syncjobservice/DiscoverySyncJobServiceCreateDTO.java:17]: why are we still using Name not id?

File: src/main/java/eu/bbmri_eric/negotiator/plugins/resourcesync/listeners/JobEventManager.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.context.ApplicationListener;
+import org.springframework.stereotype.Component;
+
+@CommonsLog
+@Component
+public class JobEventManager implements ApplicationListener<DiscoveryServiceSynchronizationEvent> {
+  @Autowired
+  private DiscoveryServiceSynchronizationJobRepository discoveryServiceSynchronizationJobRepository;
+
+  @Autowired private DiscoveryServiceClient bbmriDirectoryServiceSyncClient;
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/plugins/resourcesync/listeners/JobEventManager.java:20]: please do not autowire specific implementations

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+  @Autowired
+  private DiscoveryServiceSynchronizationJobRepository discoveryServiceSynchronizationJobRepository;
+
+  private DiscoveryServiceSynchronizationJob jobRecord;
+
+  private WebClient webClient;
+
+  // public BBMRIDiscoveryServiceClientImpl(String molgenisServiceURL) {
+  //   this.webClient = WebClient.create(molgenisServiceURL);
+  // }
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java:62]: please remove commented code

File: src/main/java/eu/bbmri_eric/negotiator/service/DiscoveryServiceClient.java [hunk 17]
```diff
@@ -0,0 +1,10 @@
+package eu.bbmri_eric.negotiator.service;
+
+public interface DiscoveryServiceClient {
+
+  public void syncAllOrganizations();
+
+  public void syncAllResources();
+
+  public void syncAllNetworks();
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/DiscoveryServiceClient.java:9]: please add javadoc

File: src/main/java/eu/bbmri_eric/negotiator/service/MolgenisServiceImplementation.java [hunk 18]
```diff
@@ -67,4 +74,71 @@ public Optional<MolgenisBiobank> findBiobankById(String id) {
...
+  //                biobank);
+  //        collections.add(collection);
+  //      }
+  //      return collections;
+  //    } catch (WebClientResponseException | WebClientRequestException e) {
+  //      log.warn(e.getMessage());
+  //      log.warn("Molgenis is not reachable!");
+  //      return Collections.emptyList();
+  //    }
+  //  }
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/MolgenisServiceImplementation.java:143]: pls remove

**Review by RadovanTomik — CHANGES_REQUESTED** [review]
please see comments and add tests

File: src/main/java/eu/bbmri_eric/negotiator/database/model/DiscoveryServiceSynchronizationJob.java (hunk 11 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/database/model/DiscoveryServiceSynchronizationJob.java:26]: Yes, this is the Entity carrying on the job statistiscs (to be improved in the second part of the PR)

File: src/main/java/eu/bbmri_eric/negotiator/database/repository/DiscoveryServiceRepository.java (hunk 13 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/database/repository/DiscoveryServiceRepository.java]: fixed in fb760484e9b2dbec0c0c5b584decfbd60305ad4d

File: src/main/java/eu/bbmri_eric/negotiator/dto/syncjobservice/DiscoverySyncJobServiceCreateDTO.java (hunk 14 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/dto/syncjobservice/DiscoverySyncJobServiceCreateDTO.java]: fixed in fb760484e9b2dbec0c0c5b584decfbd60305ad4d

File: src/main/java/eu/bbmri_eric/negotiator/plugins/resourcesync/listeners/JobEventManager.java (hunk 15 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/plugins/resourcesync/listeners/JobEventManager.java]: fixed in fb760484e9b2dbec0c0c5b584decfbd60305ad4d

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java (hunk 16 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java]: fixed in fb760484e9b2dbec0c0c5b584decfbd60305ad4d

File: src/main/java/eu/bbmri_eric/negotiator/service/DiscoveryServiceClient.java (hunk 17 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/DiscoveryServiceClient.java]: fixed in fb760484e9b2dbec0c0c5b584decfbd60305ad4d

File: src/main/java/eu/bbmri_eric/negotiator/service/MolgenisServiceImplementation.java (hunk 18 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/MolgenisServiceImplementation.java]: fixed in fb760484e9b2dbec0c0c5b584decfbd60305ad4d

File: src/main/java/eu/bbmri_eric/negotiator/plugins/resourcesync/schedulers/JobEventScheduler.java [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+@Component
+@CommonsLog
+public class JobEventScheduler {
+
+  @Autowired DiscoverySynchronizationJobService discoverySynchronizationJobService;
+
+  @Value("${synchronization.discoveryServiceId}")
+  private String synchronizationDiscoveryServiceId;
+
+  @Scheduled(cron = "${synchronization.frequency}") // Cron expression for running every minute
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/plugins/resourcesync/schedulers/JobEventScheduler.java:19]: this is why I'm not too fond of code comments: the comment is clearly outdated and therefore confusing, making the code even less readable than if the comment wasn't there

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+import org.springframework.web.reactive.function.client.WebClient;
+import org.springframework.web.reactive.function.client.WebClientRequestException;
+import org.springframework.web.reactive.function.client.WebClientResponseException;
+
+@CommonsLog
+@Service(value = "DefaultDiscoveryServiceClient")
+public class BBMRIDiscoveryServiceClientImpl implements DiscoveryServiceClient {
+
+  // @Value("${negotiator.molgenis-url}")
+  // private String molgenisURL;
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java:41]: please remove

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+  @Autowired private ResourceRepository resourceRepository;
+
+  @Autowired
+  private DiscoveryServiceSynchronizationJobRepository discoveryServiceSynchronizationJobRepository;
+
+  private DiscoveryServiceSynchronizationJob jobRecord;
+
+  private WebClient webClient;
+
+  @Autowired
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java:59]: please remove all autowires

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java [hunk 22]
```diff
@@ -0,0 +1,244 @@
...
+  public BBMRIDiscoveryServiceClientImpl(@Value("${negotiator.molgenis-url}") String molgenisURL) {
+    final int size = 16 * 1024 * 1024;
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java:61]: size of what?

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java [hunk 23]
```diff
@@ -0,0 +1,244 @@
...
+        discoveryServiceRepository.findById(Long.valueOf("1")).get();
+    AccessForm accessForm = accessFormRepository.findById(Long.valueOf("1")).get();
+    for (MolgenisCollection coll : directoryCollections) {
+      Optional<Resource> resource = resourceRepository.findBySourceId(coll.getId());
+      if (resource.isEmpty()) {
+        addMissingResource(coll, discoveryService, accessForm);
+      } else {
+        log.debug(
+            String.format("Collection %s already present, check for updates...", coll.getId()));
+        if (!resource.get().getName().equals(coll.getName())
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java:108]: why are all these methods public? The interface has way less methods

File: src/main/resources/application-dev.yaml [hunk 24]
```diff
@@ -82,4 +82,6 @@ logging:
 reminder:
   trigger-duration-days: "PT2M"
   cron-schedule-expression: "0 */5 * ? * *"
-
+synchronization:
+  frequency: "0 0 0 * * ?"
+  discoveryServiceId: "1"
```
Comment (RadovanTomik) [inline src/main/resources/application-dev.yaml:87]: why not put it in the application.yml? If it is the same everywhere

File: src/test/java/eu/bbmri_eric/negotiator/integration/plugins/resourcessync/schedulers/JobEventSchedulerTest.java [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+    Thread.sleep(5000);
+    assertEquals(1, testJobRepository.findAll().size());
+  }
+
+  @Test
+  void testJobTriggeredByChron() throws InterruptedException {
+
+    Thread.sleep(65000);
+    assertEquals(1, testJobRepository.findAll().size());
+  }
```
Comment (RadovanTomik) [inline src/test/java/eu/bbmri_eric/negotiator/integration/plugins/resourcessync/schedulers/JobEventSchedulerTest.java:54]: aren't these sleep periods too long?

**Review by RadovanTomik — CHANGES_REQUESTED** [review]

File: src/main/java/eu/bbmri_eric/negotiator/plugins/resourcesync/schedulers/JobEventScheduler.java (hunk 19 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/plugins/resourcesync/schedulers/JobEventScheduler.java]: fixed in 53131252c6f43525c044b84ec3b7c262b19835ed

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java (hunk 20 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java]: Fixed in 53131252c6f43525c044b84ec3b7c262b19835ed

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java (hunk 21 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java]: There was one not used but why all the others? They are needed as usual in Services

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java (hunk 22 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java]: Fixed in 53131252c6f43525c044b84ec3b7c262b19835ed. It is the size of the buffer limit, that had to be increased in order to properly get all collections

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java (hunk 23 above)
Comment (asulis) [reply src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java]: Fixed in 53131252c6f43525c044b84ec3b7c262b19835ed. Was a big issue, now they are all private and also the tests have been refactored accordingly (they now call only the exposed methods)

File: src/main/resources/application-dev.yaml (hunk 24 above)
Comment (asulis) [reply src/main/resources/application-dev.yaml:87]: Yes I could; should I keep it only there and in the test one?

Comment (RadovanTomik) [reply src/main/resources/application-dev.yaml:87]: if it is same everywhere then just put it in the application.yml

File: src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java [hunk 26]
```diff
@@ -0,0 +1,241 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.web.reactive.function.client.ExchangeStrategies;
+import org.springframework.web.reactive.function.client.WebClient;
+import org.springframework.web.reactive.function.client.WebClientRequestException;
+import org.springframework.web.reactive.function.client.WebClientResponseException;
+
+@CommonsLog
+@Service(value = "DefaultDiscoveryServiceClient")
+public class BBMRIDiscoveryServiceClientImpl implements DiscoveryServiceClient {
+  private BBMRIDiscoveryServiceClientImpl bbmriService = null;
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/BBMRIDiscoveryServiceClientImpl.java:39]: why is this needed?

**Review by RadovanTomik — APPROVED** [review]
